warehouse/forklift/legacy.py

```diff
@@ -607,6 +609,24 @@ def file_upload(request):
             )
         raise _exc_with_message(HTTPForbidden, msg)
 
+    # If this is a user identity (i.e: API token) but there exists
+    # a trusted publisher for this project, send an email warning that an
+    # API token was used to upload a project where Trusted Publishing is configured.
+    # Only do this once per (API token, project) combination.
+    if request.user and project.oidc_publishers:
+        macaroon: Macaroon = request.identity.macaroon
+        # If we haven't warned about use of this particular API token and project
+        # combination, send the warning email
+        if macaroon not in project.macaroons_tp_warning:
```

Comment:
I think the `macaroons_tp_warning` field is probably unnecessary on the `Project`, because we can just do something like this here:

```python
    warning_exists = request.db.query(
        exists().where(
            ProjectMacaroonWarningAssociation.macaroon_id == macaroon.id,
            ProjectMacaroonWarningAssociation.project_id == project.id,
        )
    ).scalar()
    
    if not warning_exists:
        ...
```
This is also a simpler/faster query, because we don't need to fetch all the possible warnings from the database for a project and iterate over them in Python.

Reply:
> Doesn't that happen in the backend anyway?

If by "backend" you mean "in the database", I don't think so.

> Would iterating over it fetch all the warnings from the database?

I think so. I believe this is where it happens: https://github.com/sqlalchemy/sqlalchemy/blob/da639af16f77118bc17bbf5cf78fe41dd1818168/lib/sqlalchemy/ext/associationproxy.py#L1484-L1489

You mentioned something like "don't we need this relationship to keep track of it", and I don't think we strictly need it, we can just add rows to the association table directly, and take advantage of the key constraints there.

Reply:
Got it, I made the change to use the table directly